pageserver/src/tenant.rs

```diff
@@ -488,34 +494,91 @@ impl WalRedoManager {
+///
+/// We need to store the list of offloaded timelines so that we can perform operations on them,
+/// like unoffloading them, or (at a later date), decide to perform flattening.
+/// This type has a much smaller memory impact than [`Timeline`], and thus we can store many
+/// more offloaded timelines than we can manage ones that aren't.
 pub struct OffloadedTimeline {
     pub tenant_shard_id: TenantShardId,
     pub timeline_id: TimelineId,
     pub ancestor_timeline_id: Option<TimelineId>,
     /// Whether to retain the branch lsn at the ancestor or not
     pub ancestor_retain_lsn: Option<Lsn>,
 
+    /// When the timeline was archived.
+    ///
+    /// Present for future flattening deliberations.
+    pub archived_at: NaiveDateTime,
+
     // TODO: once we persist offloaded state, make this lazily constructed
     pub remote_client: Arc<RemoteTimelineClient>,
 
     /// Prevent two tasks from deleting the timeline at the same time. If held, the
     /// timeline is being deleted. If 'true', the timeline has already been deleted.
-    pub delete_progress: Arc<tokio::sync::Mutex<DeleteTimelineFlow>>,
+    pub delete_progress: TimelineDeleteProgress,
 }
 
 impl OffloadedTimeline {
-    fn from_timeline(timeline: &Timeline) -> Self {
+    /// Obtains an offloaded timeline from a given timeline object.
+    ///
+    /// Returns `None` if the `archived_at` flag couldn't be obtained, i.e.
+    /// the timeline is not in a stopped state.
+    /// Panics if the timeline is not archived.
+    fn from_timeline(timeline: &Timeline) -> Option<Self> {
         let ancestor_retain_lsn = timeline
             .get_ancestor_timeline_id()
             .map(|_timeline_id| timeline.get_ancestor_lsn());
-        Self {
+        let archived_at = timeline
+            .remote_client
+            .archived_at_stopped_queue()?
+            .expect("must be called on an archived timeline");
+        Some(Self {
             tenant_shard_id: timeline.tenant_shard_id,
             timeline_id: timeline.timeline_id,
             ancestor_timeline_id: timeline.get_ancestor_timeline_id(),
             ancestor_retain_lsn,
+            archived_at,
 
             remote_client: timeline.remote_client.clone(),
             delete_progress: timeline.delete_progress.clone(),
+        })
+    }
+    fn from_manifest(
+        tenant_shard_id: TenantShardId,
+        remote_timeline_client: Arc<RemoteTimelineClient>,
+        manifest: &OffloadedTimelineManifest,
+    ) -> Self {
+        let OffloadedTimelineManifest {
+            timeline_id,
+            ancestor_timeline_id,
+            ancestor_retain_lsn,
+            archived_at,
+        } = *manifest;
+        Self {
+            tenant_shard_id,
+            timeline_id,
+            ancestor_timeline_id,
+            ancestor_retain_lsn,
+            archived_at,
+            remote_client: remote_timeline_client,
+            delete_progress: Timeline::make_delete_progress(),
```

Comment:
Yeah the persistent mechanism is the `deleted` flag in the timeline's json. We set it no matter whether the timeline is offloaded or not, and during tenant attach we at least resume started deletions for timelines that haven't been offloaded. For offloaded ones, we don't read the timeline's json, so we will basically wait until unoffloading.

In theory, deletions should nowadays be retried on the top level already, which is the direction John wanted to push towards. So I didn't put too much care in making this too robust (say via adding a `deleted` flag to the manifest as well).

Which reminds me, actually that's something that needs fixing. Currently we just log an info log line and don't unoffload a timeline with a set `deleted` flag. Instead we should resume deletion. But we should still log a warning, as we might have had this partially deleted zombie for quite a while, costing us money: ideally deletions are retried at the top level until they are done.

Reply:
> Which reminds me, actually that's something that needs fixing. Currently we just log an info log line and don't unoffload a timeline with a set deleted flag. Instead we should resume deletion. 

Added an item to the epic in #8088 . this can be a followup.